### PR TITLE
Validate: get line numbers when calling validate on ignition structs

### PR DESCRIPTION
**Pull request by ajeddeloh**
This makes the converter functions also modify the yaml tree, converting it into a tree that looks like it was parsed from a json source. Validate() can then be used instead of ValidateWithoutSource(), exposing locations of errors.

Depends on https://github.com/coreos/ignition/pull/355

### CODE REVIEW

File: config/astyaml/astyaml.go [hunk 1]
```diff
@@ -12,79 +12,127 @@
...
 
-func (n yamlNode) Tag() string {
-	return "yaml"
+// ChangeKey replaces the oldkey with a new key/value pair. Useful for patching
+// up a tree parsed from yaml but then used for validating an ignition structure
+func (n *YamlNode) ChangeKey(oldKeyName, newKeyName string, newValue YamlNode) error {
+	if n.Kind != yaml.MappingNode {
+		return ErrNotMappingNode
+	}
+	for i := 0; i < len(n.Children); i += 2 {
```
Comment (ajeddeloh) [inline config/astyaml/astyaml.go]: go-yaml parses a mapping type into an array of interleaved key and value nodes, so even nodes are the keys and odd nodes the values.

File: config/types/config.go [hunk 2]
```diff
@@ -49,36 +50,48 @@ type ConfigReference struct {
...
+	register2_0(func(in Config, ast validate.AstNode, out ignTypes.Config, platform string) (ignTypes.Config, report.Report, validate.AstNode) {
+		r := report.Report{}
+		cfgNode, _ := getPath(ast, "ignition", "config", "append")
+		for i, ref := range in.Ignition.Config.Append {
+			tmp, _ := get(cfgNode, i)
+			newRef, convertReport := convertConfigReference(ref, tmp)
+			r.Merge(convertReport)
+			if convertReport.IsFatal() {
+				// don't add to the output if invalid
+				continue
```
Comment (ajeddeloh) [inline config/types/config.go]: We do. It gets added to the report that gets returned. It can continue trying to parse of sections of the config and catch multiple errors at once instead of failing at the first one.

File: config/astyaml/astyaml.go [hunk 3]
```diff
@@ -12,79 +12,127 @@
...
+		words[i+1] = strings.Title(word)
+	}
+	return strings.Join(words, "")
+}
+
+func (n YamlNode) Tag() string {
+	return n.tag
+}
+
+// ChangeTreeTag changes the value Tag() returns so it
```
Comment (ajeddeloh) [inline config/astyaml/astyaml.go]: Oops, yeah the comment is incomplete. I'll change it to have a pointer receiver and modify the node itself.

**Comment by ajeddeloh** [comment]
It does feel a little hacky, but I'm not sure a good way around it. You can take out the parts the modify the tree and just leave the changing of tag to "json" and you retain line numbers for all parts that are identical but lose line number accuracy for the parts that are subtly different (it'll report the line number of the last node that was the same).

**Comment by dgonyeo** [comment]
ct [moved to using glide](https://github.com/coreos/container-linux-config-transpiler/pull/63), so you'll need to rebase and redo the vendor work with glide. If you need glide help lmk.

**Comment by dghubble** [comment]
Sorry to complicate your PR. You can go get glide github.com/Masterminds/glide, edit `glide.yaml` to declare any new dependencies or pinned versions (if any), and run `make vendor`.

**Comment by ajeddeloh** [comment]
Fixed the issue the ChangeTreeTag and rebased.

glide.yaml currently uses my fork of ignition for a needed change (https://github.com/coreos/ignition/pull/355) and once that is merged I can change that back.

File: config/astyaml/astyaml.go [hunk 4]
```diff
@@ -12,79 +12,128 @@
...
+func (n *YamlNode) ChangeKey(oldKeyName, newKeyName string, newValue YamlNode) error {
+	if n.Kind != yaml.MappingNode {
+		return ErrNotMappingNode
+	}
+	for i := 0; i < len(n.Children); i += 2 {
+		key := n.Children[i]
+		if key.Value == oldKeyName {
+			//key.Value = newKeyName
+			(*n.Children[i]).Value = newKeyName
+			*n.Children[i+1] = newValue.Node
```
Comment (dgonyeo) [inline config/astyaml/astyaml.go]: So based on your earlier comment, `n.Children` is _guaranteed_ to have an even length? Just want to make sure this won't panic.

File: config/astyaml/astyaml.go [hunk 5]
```diff
@@ -12,79 +12,128 @@
...
+	return n.tag
+}
+
+// ChangeTreeTag changes the value Tag() returns to newTag
+func (n *YamlNode) ChangeTreeTag(newTag string) {
+	/*return YamlNode{
+		key:  n.key,
+		tag:  newTag,
+		Node: n.Node,
+	}*/
```
Comment (dgonyeo) [inline config/astyaml/astyaml.go]: I take it this comment can be deleted?

File: config/types/common.go [hunk 6]
```diff
@@ -114,3 +119,41 @@ func getArgs(format, tagName string, e interface{}) []string {
 func getCliArgs(e interface{}) []string {
 	return getArgs("--%s=%v", "cli", e)
 }
+
+// Get returns the value for key, where key is an int or string and n is a
+// sequence node or mapping node, respectively.
+func get(n validate.AstNode, key interface{}) (validate.AstNode, error) {
```
Comment (dgonyeo) [inline config/types/common.go]: I'd prefer this to be named something a little more descriptive, since there are other functions here like `getCliArgs` that are entirely unrelated to `get`.

File: config/types/converter.go [hunk 7]
```diff
@@ -17,20 +17,29 @@ package types
...
 
 func register2_0(f converterFor2_0) {
 	convertersFor2_0 = append(convertersFor2_0, f)
 }
 
-func ConvertAs2_0(in Config, platform string) (ignTypes.Config, report.Report) {
+func ConvertAs2_0(in Config, platform string, ast validate.AstNode) (ignTypes.Config, report.Report) {
+	// convert our tree from having yaml tags to having json tags, so when Validate() is
+	// called on the tree, it can find the keys in the ignition structs (which are denoted
+	// by `json` tags
```
Comment (dgonyeo) [inline config/types/converter.go]: )

File: config/astyaml/astyaml.go (hunk 4 above)
Comment (ajeddeloh) [inline config/astyaml/astyaml.go]: Yes. See https://github.com/ajeddeloh/yaml/blob/public-node/decode.go#L175 The key/values are always added in pairs.

**Comment by ajeddeloh** [comment]
Added unit tests.

**Comment by ajeddeloh** [comment]
Ignition vendoring done.

**Comment by dgonyeo** [comment]
can you rebase?

**Comment by ajeddeloh** [comment]
rebased
